final-project/application.py

```diff
@@ -486,6 +486,84 @@ def search_category():
         users_selected_category=users_selected_category)
 
 
+# Store references to product research
+@app.route("/reference")
+def reference():
+    """Search for references to product research."""
+
+    # Keep the user logged in
+    user_id = session.get("user_id")
+
+    # Get the product name requested by the users
+    product_to_provide_references_for = request.args.get(
+            "product")
+
+    # If no name is given, then ask the user for one.
+    if not product_to_provide_references_for:
+        return apology("Please specify product")
+    elif product_to_provide_references_for == None:
+        return apology("Please specify product")
+
+    # Search for references to a given product
+    references_for_product = db.execute("""
+            SELECT research.link, title
+            FROM research
+            LEFT JOIN products
+            ON products.id=research.product_id
+            WHERE products.product_name=:product
+            """, product=product_to_provide_references_for)
+
+    # Check whether we found references
+    if not references_for_product:
+        return apology("We have no references. Please help us research")
+    elif references_for_product == None:
+        return apology("We have no references. Please help us research")
+
+    # Save the number of research articles for this products.
+    number_of_references = len(references_for_product)
+
+    # Save the titles and links to each reference.
+    # Create an empty list to save our titles.
+    reference_titles = []
+    # Create an empty list to save our links.
+    reference_links = []
+
+    # Loop through each title in our list of references
+    for reference in range(number_of_references):
+
+        # Ensure we have a title.
+        if not references_for_product[reference]["title"]:
+            # If there is no title, then use the link.
+            reference_titles.append(references_for_product[reference]["link"])
+        # If there is a title, then use the title.
+        else:
+            # Save the title.
+            reference_titles.append(references_for_product[reference]["title"])
+
+        # Save the link.
+        reference_links.append(references_for_product[reference]["link"])
```

Comment:
This works fine, though I wonder if you also like to have titles and links associated w/ each other? In that case, I would look at tuples or other data structures to keep the data together.

Also take a look at http://book.pythontips.com/en/latest/map_filter.html to make your python more functional and declarative. They are easier to read, less state-dependent, and easier to debug! Woo for functional programming!

For future-future reference, something like this is also fun 😄  https://github.com/ReactiveX/RxPY

Reply:
The information is already in a list of dictionary objects, which makes it pretty easy to access. However, I wasn't sure if I should store the information into variables for readability. When do you create variables versus just keep everything in a more complex data structure?

For things like `request.args.get(something_from_the_get_request)`, would this information immediately get stored into a variable for reuse? Then for things like `reference[0]["title"]` and `reference[0]["link"]`, do I keep the information in the list of dictionary objects?